### ISSUE TITLE
Potential issue with TOC2 and Material Files

### ISSUE THREAD

**Issue by bigboipete** [zhanghai/MaterialFiles#517]
Hi there,

There is an issue in TOC2 metronome app which might relate to Material Files.

See [this](https://github.com/thetwom/toc2/issues/35#issuecomment-850806459) comment by @thetwom.

I just want to draw your attention to it and maybe you can connect to find out, what's it about.

**Comment by thetwom** [comment]
Maybe I should describe the issue in more detail:
I am sending a ACTION_DOCUMENT_OPEN-Intent via an ActrivityResultContract to open a file picker. This works fine so far. But after picking a file, the ActivityResultContract doesn't return an uri to the file but null. 

All works fine when MaterialFiles is not installed. But as soon as it is installed I don't get an uri. 

Strange thing is, that it also fails to work, when I am NOT using MaterialFiles as file picker, but when MaterialFiles is installed.

In summary, an installed MaterialFiles prevents picking files ...

**Comment by zhanghai** [comment]
I guess it could be becuase there is an additional chooser activity when there's more than one app handling `ACTION_OPEN_DOCUMENT`, and it isn't fowarding the result to the caller correctly in some corner cases.

FWIW, Android 11+ has restricted the handler for that intent to the DocumentsUI system app by adding `android:priority="100"`, so having Material Files installed or not shouldn't matter. But still interestingly, when I tested locally on my OnePlus OxygenOS Android 9, file picking via the system file picker in most apps still worked fine for me. Maybe an OEM specific issue?

**Comment by thetwom** [comment]
Thanks for taking the time and answering to this issue!  Here my current observations:
1. I tested this on Lineage 17.1 (android 11), but the problem was reported by @bigboipete on ShiftOs (AOSP 8.0.0). However, I can't currently say, that uninstalling MaterialFiles on ShiftOS helps. I am personally just able to test this on Lineage.
2. Things work fine, if I don't use MaterialFiles as FilePicker, but open MaterialFiles manually and ask it to "open file in ..."

My initial guess was that MaterialFiles isn't passing the permission to read the file ... and maybe other apps have requested file reading permissions independently? But it still would not explain, why the SystemPicker doesn't work when MaterialFiles is installed. 

If you have no quick idea what would cause this issue, I will see that I provide an isolated test. And to be honest, currently I can't "promise", that it not an issue of toc2 app.

**Comment by thetwom** [comment]
One year later I finally took the time and investigated a bit further. I seems, that if an app uses "singleInstance" as launch-mode, the returned uri will be null. The problem is solved when e.g. using "singleTask". Here is a very simple test app, which demonstrates this: [FilePickerTest](https://github.com/thetwom/FilePickerTest). One can click the button and try to open a text file. It doesn't open. However, if one goes to the manifest and changes singleInstance to singleTask, everything is fine. All important code is in the MainActivity. The issue seems pretty consistent and appears on different android versions and flavours. 

For me, I guess, this finally solves the issue, still I wonder, why the problem only occurs when MaterialFiles is installed. 

So, feel free to close this issue or not, ...

**Comment by zhanghai** [comment]
Sorry I'm still confused, because this app doesn't have any activity with singleTask or singleInstance launch mode. Or did you mean the issue is actually unrelated to Material Files?

**Comment by thetwom** [comment]
Sorry for the confusion. Let me describe the issue more formally ...

Steps to reproduce:
1. Clone, compile and run the test app [FilePickerTest](https://github.com/thetwom/FilePickerTest)
2. Run the app and click the button to load a text file
3. FilePicker is shown ... choose some text file
4. The app appears again and ideally shows the content of the text file

Expected behavior:
The app opens the text file and shows the content

Actual behaviors:
- If MaterialFiles is NOT installed: Text file is correctly opened
- If MaterialFiles IS installed: App shows "Failed to load file" (reason is that the app gets "null" as uri). It actually doesn't matter if we use the MaterialFiles-FilePicker or the system file picker. As soon as MaterialFiles is installed, we ALWAYS get null.

Ways to circumvent the problem:
- In the test app, open the Manifest and change the launch mode from "SingleInstance" to "SingleTask". 

What I don't know is who is to "blame":
- The test app, since it is a bad idea to use launch-mode SingleInstance?
- Android itself, since it shows unpredictable behavior for SingleInstance-apps?
- MaterialFiles, since the issue occurs only when it is installed?

Since I don't if MaterialFiles is really the reason for the problem feel free to close this issue. Thanks for taking the time to answer the questions and thanks for all your efforts for developing the app.

**Comment by zhanghai** [comment]
I see, thanks. I think this is more likely a platform bug.
